### ISSUE TITLE
Uncaught TypeError: fn is not a function when modify shared-routing dashboard public index.html

### ISSUE THREAD

**Issue by lidianhao123** [module-federation/module-federation-examples#322]
I find http://localhost:3000/remoteEntry.js script file is load twice, one in head and another in body. So I delete the script in head. 
Modify code [module-federation-examples/shared-routing/dashboard/public/index.html](https://github.com/module-federation/module-federation-examples/blob/master/shared-routing/dashboard/public/index.html)
```html
<html>

<head>
  <script src="http://localhost:3000/remoteEntry.js"></script>
  <!-- <script src="http://localhost:3001/remoteEntry.js"></script> -->
  <script src="http://localhost:3002/remoteEntry.js"></script>
  <script src="http://localhost:3003/remoteEntry.js"></script>
  <script src="http://localhost:3004/remoteEntry.js"></script>
</head>

<body>
  <div id="root"></div>
</body>

</html>
```

then run the Apps,  visit http://localhost:3001/dashboard
![image](https://user-images.githubusercontent.com/7970645/90842923-09109200-e393-11ea-9d39-40c4a6eb127a.png)

index.html structure
![image](https://user-images.githubusercontent.com/7970645/90843007-33fae600-e393-11ea-87fb-f33b3325b2a2.png)

Why 3001/remoteEntry.js file load order will cause the error? must manual add script in order?

**Comment by ScriptedAlchemy** [comment]
The remote is booting after its host, which likely messes with the resolution. Most likely, html-webpack-plugin is not excluding the remote and emitting its own. You could also have webpack load these remotes on its own with this syntax, then webpack manages script loading https://github.com/module-federation/module-federation-examples/blob/master/bi-directional/app1/webpack.config.js#L32

**Comment by vania22** [comment]
The problem I think is because you have the element with id which is equals to the name of your remote component. Try to find that element and change the id of it

**Comment by jasveer1997** [comment]
@ScriptedAlchemy @vania22 I wanted to load the remote entry at the initial load in html template, however I am getting the attached error. I have checked for duplicate element id (to federation_app_name) or order of main.js (inside <body/>) or remoteEntry(inside <head/>)
It was working fine with `'local_name': federation_app_name@${ENV.FEDERATION_APP_URL}`
![image](https://user-images.githubusercontent.com/52408834/111510532-e2bda400-8773-11eb-91fa-32856bb5fdb1.png)
![image](https://user-images.githubusercontent.com/52408834/111510882-34662e80-8774-11eb-915e-2aa8e41747ae.png)
![image](https://user-images.githubusercontent.com/52408834/111511477-ca01be00-8774-11eb-8704-76d943d98eab.png)

**Comment by vadymrybak** [comment]
I also have a issue related to "fn is not a function". In my scenario I have:

- "common" react app that only exposes Header and Footer components 
```
new ModuleFederationPlugin({
            name: "common",
            library: { type: "var", name: "common" },
            filename: "remoteEntry.js",
            exposes: {
                // expose each component
                "./Header": "./src/Components/Header",
                "./Footer": "./src/Components/Footer"
            },
            shared: ["react", "react-dom", "react-router-dom", "react-bootstrap", "bootstrap", "react-icons"],
        }),
```
- "workflow" react app that only exposes some SVG workflow diagram component
```
new ModuleFederationPlugin({
            name: "workflow",
            library: { type: "var", name: "workflow" },
            filename: "remoteEntry.js",
            exposes: {
                // expose each component
                "./Workflow": "./src/Components/Workflow"
            },
            shared: ["react", "react-dom", "react-router-dom"],
        }),
```
- "host" react app that consumes header and footer from "common" and workflow from "workflow"

Then I did a test: 
1. I created "host2" react app and made "host" app expose its entire App component with imported header, footer, workflow.  
```
 new ModuleFederationPlugin({
                name: "host2",
                library: { type: "var", name: "host2" },
                remotes: {
                    host: "host"
                },
                shared: ["react", "react-dom", "react-router-dom", "react-bootstrap", "bootstrap", "react-icons"],
            }),
```
2. So "host" is now bi-directional: it's consumes(remotes) common and workflow as well as exposes "HostApp" component.
```
new ModuleFederationPlugin({
                name: "host",
                library: { type: "var", name: "host" },
                remotes: {
                    common: "common",
                    workflow: "workflow"
                },
                exposes: {
                    "./HostApp": "./src/Components/App",
                },
                shared: ["react", "react-dom", "react-router-dom", "react-bootstrap", "bootstrap", "react-icons"],
            }),
```
Host2 app now only consumes HostApp from "host" and I only provided remoteEntry script reference to "host" app. 
In this case I'm getting "fn is not a function" error when trying to load "host2" app. 
Only when I add additional script tags pointing to remoteEntry of "common" and "workflow", this error goes away. 

Basically, my question is why my "host2" app needs to know about remoteEntries in "common" and "workflow"?

**Comment by Matin297** [comment]
> The problem I think is because you have the element with id which is equals to the name of your remote component. Try to find that element and change the id of it

Thanks. It just solved my problem!

**Comment by debo07** [comment]
> The problem I think is because you have the element with id which is equals to the name of your remote component. Try to find that element and change the id of it

@vania22 Thanks, this saved my day. But could you explain the reason behind it! I am new to this module-federation concept, why can't I keep element id same as any remote component name?

**Comment by vsramanujan** [comment]
My suspicion is that the webpack MF plugin creates global variables with the name of the remotes (in the remoteEntry file) and these are then used to dynamically load the remote module. 
Browsers also create global variables for all elements that have an id defined. This possibly overwrites the variable that the plugin sets, hence causing the error.

**Comment by adamkovacs-dt** [comment]
For everyone who arrives here, please check your global variables in the console, and the name of your remotes for collisions. For us the problem was that we called our remote `navigator` but we had a global variable with that name already defined.

**Comment by mdipanjan** [comment]
> The problem I think is because you have the element with id which is equals to the name of your remote component. Try to find that element and change the id of it
@vania22 
Thank you. This solved my issue 💯

**Comment by Paniraj2010** [comment]
Hi, I have the same issue. 

I am trying to integrate React MFE in angular host. where in React MFE expecting pass props. it all works good when react MFE access in react host. But when I Try to access in Angular host I am getting **fn is not a function while loading**  

**Also React MFE expecting input parameter it was implemented using Mount.** 

I have implemented in Angular using a wrapper component by referring **https://www.angulararchitects.io/en/aktuelles/multi-framework-and-version-micro-frontends-with-module-federation-the-good-the-bad-the-ugly/** however when I run the application it gives me **fn is not a function while loading**


```
import { mount } from `'mfeHierarchy/MfeHierarchy';`
Component({
  template: `
  <div #placeHolder></div>
`,
})
export class WrapperComponent implements AfterContentInit {

  @ViewChild('placeHolder', { read: ViewContainerRef, static: true })
    viewContainer!: ViewContainerRef;
    
  userContext:any;
 
  constructor(private route: ActivatedRoute) {
    const queryParams = new URLSearchParams(window.location.search);
    this.userContext = {
      userName: queryParams.get("NTLogin"),
      userEmail: queryParams.get("Email")
    };
   }
  
  ngAfterContentInit(): void {
    
    console.log('with in ngAfterContentInit');
   
    console.log(mount.apid);
}

```

**In the above code mount is imported from the remote module. As soon as I access Mount with in ngAfterContentInit() it is throwing error fn is not a function while loading**

I see following comment "The remote is booting after its host, which likely messes with the resolution." from https://github.com/module-federation/module-federation-examples/issues/322 but  I am not sure how can we make sure Remote MFE is booting before HOST.

Please assist.

@ScriptedAlchemy

**Comment by Paniraj2010** [comment]
Hello, Angular 13 with webpack 5 causing this issue, once I update angular back to angular 12 everything started working.

Webpack 5 with Angular 13 has issue with remotes loading. **I am getting issue with below lines, which get auto generate with webpack run time** 

Due to this application was failing to load the remote module. Please fix this in the webpack issue with Angular 13

```
var handleFunction = (fn, arg1, arg2, d, next, first) => {
				try {
					**var promise = fn(arg1, arg2);**
					if(promise && promise.then) {
						var p = promise.then((result) => (next(result, d)), onError);
						if(first) promises.push(data.p = p); else return p;
					} else {
						return next(promise, d, first);
					}
				} catch(error) {
					onError(error);
				}
			}
```

**Comment by Hall-e3** [comment]
make the id of the div in the body of the HTML file of the different micro apps i.e container and dashboard different

**Comment by muhammadmohie** [comment]
I think this may solve the problem.

I replaced the **id** attribute with a **class** attribute and worked for me.

### For example:-

**In the index.html files**:
`<div class="dashboard"></div>`

**In the index.js file**:
`document.querySelector(".dashboard").innerHTML = dashboard;`

**Comment by msgtobala** [comment]
> The problem I think is because you have the element with id which is equals to the name of your remote component. Try to find that element and change the id of it

@vania22  Thanks for the answer. Can you please explain why is that happened?
